src/java.base/share/native/libfallbackLinker/fallbackLinker.c

```diff
@@ -0,0 +1,203 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "jdk_internal_foreign_abi_fallback_LibFallback.h"
+
+#include <ffi.h>
+
+#include <errno.h>
+#ifdef _WIN64
+#include <Windows.h>
+#include <Winsock2.h>
+#endif
+
+#include "jlong.h"
+
+static JavaVM* VM;
+static jclass LibFallback_class;
+static jmethodID LibFallback_doUpcall_ID;
+static const char* LibFallback_doUpcall_sig = "(JJLjava/lang/invoke/MethodHandle;)V";
+
+JNIEXPORT void JNICALL
+Java_jdk_internal_foreign_abi_fallback_LibFallback_init(JNIEnv* env, jclass cls) {
+  (*env)->GetJavaVM(env, &VM);
+  LibFallback_class = (*env)->FindClass(env, "jdk/internal/foreign/abi/fallback/LibFallback");
+  LibFallback_doUpcall_ID = (*env)->GetStaticMethodID(env,
+    LibFallback_class, "doUpcall", LibFallback_doUpcall_sig);
+}
+
+JNIEXPORT jlong JNICALL
+Java_jdk_internal_foreign_abi_fallback_LibFallback_sizeofCif(JNIEnv* env, jclass cls) {
+  return sizeof(ffi_cif);
+}
+
+JNIEXPORT jint JNICALL
+Java_jdk_internal_foreign_abi_fallback_LibFallback_ffi_1prep_1cif(JNIEnv* env, jclass cls, jlong cif, jint abi, jint nargs, jlong rtype, jlong atypes) {
+  return ffi_prep_cif(jlong_to_ptr(cif), (ffi_abi) abi, (unsigned int) nargs, jlong_to_ptr(rtype), jlong_to_ptr(atypes));
+}
+JNIEXPORT jint JNICALL
+Java_jdk_internal_foreign_abi_fallback_LibFallback_ffi_1get_1struct_1offsets(JNIEnv* env, jclass cls, jint abi, jlong type, jlong offsets) {
+  return ffi_get_struct_offsets((ffi_abi) abi, jlong_to_ptr(type), jlong_to_ptr(offsets));
+}
+
+static void do_capture_state(int32_t* value_ptr, int captured_state_mask) {
+    // keep in synch with jdk.internal.foreign.abi.CapturableState
+  enum PreservableValues {
+    NONE = 0,
+    GET_LAST_ERROR = 1,
+    WSA_GET_LAST_ERROR = 1 << 1,
+    ERRNO = 1 << 2
+  };
+#ifdef _WIN64
+  if (captured_state_mask & GET_LAST_ERROR) {
+    *value_ptr = GetLastError();
+    value_ptr++;
+  }
+  if (captured_state_mask & WSA_GET_LAST_ERROR) {
+    *value_ptr = WSAGetLastError();
+    value_ptr++;
+  }
+#endif
```

Comment:
This looks incorrect. Look at the changes to `downcallLinker.cpp`, the `value_ptr++` should be moved outside of the if statements. (it only matters on Windows though, which we can't test at the moment for the fallback linker, but let's keep the 2 impls in sync for now)